src/asymmetric_encryption.cpp

```diff
@@ -0,0 +1,231 @@
+    return *this;
+}
+
+std::ostream& AsymmetricEncryption::CryptoData::operator<<(std::ostream& os)
+{
+    os << getString();
+    return os;
+}
+
+std::string AsymmetricEncryption::CryptoData::getString() const {
+    return std::string(_data.begin(), _data.end());
+}
+
+std::vector<uint8_t> AsymmetricEncryption::CryptoData::getByteArray() const {
+    return _data;
+}
+
+AsymmetricEncryption::CryptoData AsymmetricEncryption::encrypt(AsymmetricPublicKey key, 
+                                                               const RSAPadding &pad,
+                                                               const CryptoData& message)
+{
+    SSL_RSA_OAEP_LABEL_Ptr label_copy(nullptr);
+    SSL_RSA_ENCRYPTION_DATA_Ptr encryptedMessage{nullptr};
+    size_t encryptedMessageLen{0};
+    int maxSize{0};
+
+    const auto keyCtx = _EVP_PKEY_CTX_new(key.internal());
+
+    try {
+        const auto paddingMode = pad.getPadding();
+
+        if (!keyCtx.get()) {
+            throw MoCOCrWException("Encryption context is empty");
+        }
+
+        switch(paddingMode) {
+            case RSAPaddingMode::PKCS1: {
+                maxSize = RSA_size(key.internal()->pkey.rsa) - c_pkcsMaxSizeSubtract;
+                break;
+            }
+            case RSAPaddingMode::OAEP: {
+                const auto oaepPaddingMode = static_cast<const OAEPPadding&>(pad);
+                maxSize = RSA_size(key.internal()->pkey.rsa) -
+                    (2 * _EVP_MD_size(_getMDPtrFromDigestType(oaepPaddingMode.getHashingFunction())) -  2);
+                break;
+            }
+            case RSAPaddingMode::NONE: {
+                maxSize = RSA_size(key.internal()->pkey.rsa);
+                break;
+            }
+            default: throw MoCOCrWException("Unsupported encryption mode");
+        }
+
+        /* Validate message size */
+        if (static_cast<int>(message.getByteArray().size()) > maxSize) {
+            throw MoCOCrWException("Message too long for RSA key size");
+        }
+        _CRYPTO_malloc_init();
+        _EVP_PKEY_encrypt_init(keyCtx.get());
+        _EVP_PKEY_CTX_set_rsa_padding(keyCtx.get(), static_cast<int>(paddingMode));
+
+        if (RSAPaddingMode::OAEP == paddingMode) {
+            const auto oaepPaddingMode = static_cast<const OAEPPadding&>(pad);
+
+            _EVP_PKEY_CTX_set_rsa_oaep_md(keyCtx.get(),
+                                          _getMDPtrFromDigestType(oaepPaddingMode.getHashingFunction()));
+
+            _EVP_PKEY_CTX_set_rsa_mgf1_md(keyCtx.get(),
+                                          _getMDPtrFromDigestType(oaepPaddingMode.getMaskingFunction()));
+
+            if(oaepPaddingMode.getLabel().size() > 0){
+
+                /* Make a copy of the label, since the context takes ownership of it when calling
+                 * '_EVP_PKEY_CTX_set_rsa_oaep_label()' function*/
+                label_copy.reset(static_cast<uint8_t*>(_OPENSSL_malloc(oaepPaddingMode.getLabel().size())));
```

Comment:
An issue due to a implicit conversion which losses precision may lead to a severe issue regarding to memory allocation and memory access.  

1. _OPENSSL_malloc receives an int, which is obtained from a implicit conversion of unsigned long to int and may result in negative number. And therefore, try to allocated negative memory. 

2. The value used for the memory allocation by the function OPENSSL_malloc may be different (due to the implicit conversion) from the value use in the memcpy which defined the amount o memory allocated/copied.

I want to highlight, that such case would only appends if  "oaepPaddingMode" size exceeds the int max value and may not be possible to use such padding Mode. 

I would like to ear your option about this topic. However, i suggest that at least a comment to be added at function header to warning for such cases.